### ISSUE TITLE
FilePath.setContent should be muti-process safe

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3694]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)| @glyph reported|
|-|-|
|Trac ID|trac#3694|
|Type|enhancement|
|Created|2009-03-20 06:46:20Z|
|Branch|https://github.com/twisted/twisted/tree/setContent-multiprocess-3694-2|

Right now `FilePath.setContent` will always name its temporary file `<filename>.new`, which means that two concurrent writers can easily stomp over each other.  It should really use `temporarySibling` to create a unique temporary filename for each writer.

Attachments:

* [TwistedTicket3694.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/cbc/cbc655760122645cedde314635041e3426c143b2/b611a69d0e0b811bc6eb5c732da7f7c377c62865.patch) (625 bytes) - added by sagen on 2010-04-27 21:03:18Z - Use temporarySibling( ) if ext is not specified

<details><summary>Searchable metadata</summary>

```
trac-id__3694 3694
type__enhancement enhancement
reporter__glyph glyph
priority__normal normal
milestone__ 
branch__branches_setContent_multiprocess_3694_2 branches/setContent-multiprocess-3694-2
branch_author__glyph glyph
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1237531580000000 1237531580000000
changetime__1274738561000000 1274738561000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|<img alt="sagen's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_1"></a>|sagen commented|
|-|-|

FYI, the TwistedTicket3694.patch breaks twisted.test.test_paths.FilePathTestCase.test_setContentFileClosing, so more needs to be done.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun commented|
|-|-|

It'd also be nice to have a test which demonstrates that the change makes `setContent` multi-process safe.  I don't have any suggestions for a simple way to write such a test.  I do know that there's a branch somewhere that does a whole lot of stuff to help make it possible to write `FilePath` unit tests, though.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_4"></a>|@glyph commented|
|-|-|

Replying to [exarkun](#note_2):
> It'd also be nice to have a test which demonstrates that the change makes `setContent` multi-process safe.  I don't have any suggestions for a simple way to write such a test.  I do know that there's a branch somewhere that does a whole lot of stuff to help make it possible to write `FilePath` unit tests, though.

I think the branch that you're referring to is linked from [#2931](https://github.com/twisted/twisted/issues/2931).

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_5"></a>|@glyph commented|
|-|-|

(In [28964]) Branching to 'setContent-multiprocess-3694'

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_6"></a>|@glyph commented|
|-|-|

[Build results.](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/setContent-multiprocess-3694)

As it turns out, I didn't even need to mock out the `os` module to test this, so I didn't bring anything in from [#2931](https://github.com/twisted/twisted/issues/2931) except a docstring.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph removed owner<br>@glyph set status to `new`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun set owner to @glyph<br>@exarkun set status to `new`|
|-|-|

  1. `TrackingFilePath` has some instance attributes that maybe should have documentation.
  1. `test_setContentNameCollision` uses `assertEqual` and `assertNotEqual`, should be `assertEquals` and `assertNotEquals`
  1. More similar in `test_setContentExtension`
  1. Should we have a ticket for implementing this more properly on versions of Windows where it's possible (maybe we have a ticket like that already)?  If so, the `setContent` docstring should refer to it I think.

The rest looks fine.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_10"></a>|@glyph commented|
|-|-|

(In [29010]) address review point 1: documentation for TrackingFilePath instance attributes.  refs [#3694](https://github.com/twisted/twisted/issues/3694)

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_11"></a>|@glyph commented|
|-|-|

(In [29011]) address review point 2: assertEqual -> assertEquals in test_setContentNameCollision.  refs [#3694](https://github.com/twisted/twisted/issues/3694)

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_12"></a>|@glyph commented|
|-|-|

(In [29014]) address review point 4: add a link to the Windows-atomic-rename ticket. refs [#3694](https://github.com/twisted/twisted/issues/3694)

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_13"></a>|@glyph set status to `closed`|
|-|-|

(In [29015]) Merge setContent-multiprocess-3694: make setContent safe for use by multiple concurrent writers

Author: glyph

Reviewer: exarkun

Fixes: [#3694](https://github.com/twisted/twisted/issues/3694)

This improves the multi-process concurrency of FilePath.setContent so that multiple concurrent writers will not stomp over each other, as they previously would have.  The temporary filename is now unpredictable.

This also adds an extension argument to temporarySibling, so temporary files can have a recognizable filename pattern, and a significantly better FilePath.setContent docstring.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_14"></a>|@glyph set status to `reopened`|
|-|-|

(In [29018]) Revert r29015

The new setContent uses temporarySibling, which calls requireCreate, which makes open() call create(), and create() doesn't call a mode argument, but in order to be correct it needs to take O_BINARY on Windows.  This is really a separate bug, but this fix should be merged after it is in.

Reopens: [#3694](https://github.com/twisted/twisted/issues/3694)

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_15"></a>|@glyph removed owner<br>@glyph set status to `new`|
|-|-|

I think we can revert the revert directly onto trunk once [#4453](https://github.com/twisted/twisted/issues/4453) is merged.  Anyone want to confirm?

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_16"></a>|@glyph commented|
|-|-|

(In [29028]) Branching to 'setContent-multiprocess-3694-2'

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_17"></a>|@glyph commented|
|-|-|

Changes are applied on a branch after [#4453](https://github.com/twisted/twisted/issues/4453)'s merge to trunk, now.  [Build results will be here](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/setContent-multiprocess-3694-2), when they're ready.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_18"></a>|@glyph commented|
|-|-|

Hooray, all the builds are green except for [ticket:4454 the known failures on Lucid].

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_19"></a>|@exarkun set owner to @glyph|
|-|-|

Looks good to re-merge (from the new branch, which was run on buildbot).

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_20"></a>|@glyph set status to `closed`|
|-|-|

(In [29034]) Merge setContent-multiprocess-3694-2: make setContent multiprocess safe

Author: glyph

Reviewer: exarkun

Fixes: [#3694](https://github.com/twisted/twisted/issues/3694)

twisted.python.filepath.FilePath.setContent is now safe for multiple processes to use concurrently.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
